OfflineService/OfflineServiceDataSource.swift

```diff
@@ -0,0 +1,142 @@
+        
+        self.delegate?.offlineServiceDataSource(self, didUpdate: [offlineDataItem])
+    }
+    
+    public func onExpired(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack) {
+        NSLog("[OfflineServiceObserver] \(#function), identifier: \(metadata.id)")
+    }
+    
+    public func onErrored(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack) {
+        NSLog("[OfflineServiceObserver] \(#function), identifier: \(metadata.id)")
+        
+        self.delegate?.offlineServiceDataSource(self, didFail: OfflineServiceError.genericError(message: "Error occured for: \(metadata.id)"))
+    }
+    
+    public func onDeleting(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack, callback: @escaping OfflineDataPackAcknowledgeCallback) {
+        NSLog("[OfflineServiceObserver] \(#function), identifier: \(metadata.id)")
+        callback()
+    }
+    
+    public func onDeleted(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata) {
+        NSLog("[OfflineServiceObserver] \(#function), identifier: \(metadata.id)")
+        
+        let offlineDataItem = OfflineDataItem(dataRegionMetadata: metadata, domain: domain)
+        self.delegate?.offlineServiceDataSource(self, didUpdate: [offlineDataItem])
+    }
+    
+    public func onInitialized() {
+        NSLog("[OfflineServiceObserver] \(#function)")
+    }
+    
+    public func onIdle() {
+        NSLog("[OfflineServiceObserver] \(#function)")
+    }
+    
+    public func onLogMessage(forMessage message: String) {
+        NSLog("[OfflineServiceObserver] \(#function): \(message)")
+    }
+    
+    public var peer: MBXPeerWrapper?
+    
+    deinit {
+        OfflineServiceManager.unregister(self)
+    }
+    
+    // MARK: - Public methods
+    
+    /**
+     Subscribe to updates from `OfflineService` and list all available regions for `username` and `accessToken` specified in `OfflineServiceConstants`.
+     */
+    public func startObservingAvailableRegions() {
+        OfflineServiceManager.register(self)
+        
+        OfflineServiceManager.instance.listAvailableRegions { [weak self] (expected) in
+            guard let self = self else { return }
+            
+            if let error = expected?.error as? OfflineDataError {
+                self.delegate?.offlineServiceDataSource(self, didFail: OfflineServiceError.genericError(message: error.message))
+
+                return
+            }
+
+            let offlineDataRegions = expected?.value as? Array<Any>
+            var offlineDataItems = [OfflineDataItem]()
+            offlineDataRegions?.forEach {
+                if let metadata = $0 as? OfflineDataRegionMetadata {
+                    offlineDataItems.append(OfflineDataItem(dataRegionMetadata: metadata))
+                }
+            }
+            
+            if !offlineDataItems.isEmpty {
```

Comment:
empty array is still a result, we may want to stop showing previously gotten regions

Reply:
> The main complaint is the hardcoded values. And I don't fully share your way for configuring navigators, not that I need fame, but there is a pr that does it better, the pr is #2593

Even though current implementation adds support to use tiles downloaded via offline service for route request, active guidance and free drive by extending public APIs I agree that PR in #2593 can address navigators configuration better and provide singe point of entry for all Navigators we currently use. My concerns can be found in https://github.com/mapbox/mapbox-navigation-ios/pull/2653#issuecomment-708842296 as well.

I encourage Core Navigation team members to review your Navigator configuration PR as well.

Reply:
Let's address it in https://github.com/mapbox/mapbox-navigation-ios/pull/2704.